src/lib/optimizer/strategy/join_to_predicate_rewrite_rule.cpp

```diff
@@ -0,0 +1,200 @@
+    const auto candidate_exp = std::dynamic_pointer_cast<BinaryPredicateExpression>(candidate->predicate());
+
+    if (!candidate_exp) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    DebugAssert(candidate_exp, "Need to get a predicate with a BinaryPredicateExpression.");
+
+    // Only predicates that filter using equals cond. on a column with a constant value are of use to our optimization
+    // these conditions have the potential (given filtered column is a UCC) to single out a maximum of one result tuple
+    if (candidate_exp->predicate_condition != PredicateCondition::Equals)
+      return LQPVisitation::VisitInputs;
+
+    auto candidate_column_expr = std::dynamic_pointer_cast<LQPColumnExpression>(candidate_exp->left_operand());
+    std::shared_ptr<ValueExpression> candidate_value_expr = nullptr;
+    if (candidate_column_expr) {
+      candidate_value_expr = std::dynamic_pointer_cast<ValueExpression>(candidate_exp->right_operand());
+    } else {
+      candidate_column_expr = std::dynamic_pointer_cast<LQPColumnExpression>(candidate_exp->right_operand());
+      candidate_value_expr = std::dynamic_pointer_cast<ValueExpression>(candidate_exp->left_operand());
+    }
+
+    // should not have a case where we run into no column, but may be a case where we have no value but compare columns
+    if (!candidate_column_expr || !candidate_value_expr) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    // if Predicate column expr != join column expression, check whether column referenced is still available in join
+    if (!expression_evaluable_on_lqp(candidate_column_expr, *removable_subtree)) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    // Check for uniqueness
+    auto testable_expressions = ExpressionUnorderedSet{};
+    testable_expressions.insert(candidate_column_expr);
+
+    if (!removable_subtree->has_matching_unique_constraint(testable_expressions)) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    valid_predicate = candidate;
+    return LQPVisitation::DoNotVisitInputs;
+  });
+
+  if (!valid_predicate) {
+    return false;
+  }
+
+  return true;
+}
+
+void JoinToPredicateRewriteRule::_perform_rewrite(const std::shared_ptr<JoinNode>& join_node,
+                                                  const LQPInputSide& removable_side,
+                                                  const std::shared_ptr<PredicateNode>& valid_predicate) const {
+  const auto param_ids = std::vector<ParameterID>{};
+  const auto param_expressions = std::vector<std::shared_ptr<AbstractExpression>>{};
+
+  if (removable_side == LQPInputSide::Left) {
+    join_node->set_left_input(join_node->right_input());
+  }
+  join_node->set_right_input(nullptr);
+
+  // get the join predicate, as we need to extract which column to filter on
+  std::shared_ptr<BinaryPredicateExpression> join_predicate = nullptr;
+  const auto& join_predicates = join_node->join_predicates();
```

Comment:
You checked that the join has one single binary üpredicate before, so add acomment and don't do that twice.